lib/util/pthread_create_blocking.c

```diff
@@ -0,0 +1,279 @@
+		warn0("pthread_mutex_unlock: %s", strerror(rc));
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err1:
+	if ((rc = pthread_mutex_unlock(U->mutex_p)))
+		warn0("pthread_mutex_unlock: %s", strerror(rc));
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * pthread_cleanup_{push,pop}() can be implemented as either functions or as
+ * macros, and must appear as a pair in the same lexical scope.  As such,
+ * we cannot have a construct such as
+ *     if (clean_func != NULL)
+ *             pthread_cleanup_push(clean_func, user_cookie);
+ * Instead, if the user has passed a NULL for clean_func, we set it to this
+ * "no-op" function.
+ */
+static void
+cleanup_fake(void * user_cookie)
+{
+
+	(void)user_cookie; /* UNUSED */
+}
+
+static void *
+wrapped_thread(void * cookie)
+{
+	struct wrapped_cookie * U = cookie;
+	void * (*run_routine)(void *) = U->run_routine;
+	void (*cleanup_routine)(void *) = U->cleanup_routine;
+	void * user_cookie = U->user_cookie;
+	void * rc;
+
+	/* Run the user's init function, if provided. */
+	if (U->init_routine) {
+		if ((*U->rc_init_p = U->init_routine(user_cookie)) != 0) {
+			if (set_initialized(U) != 0)
+				*U->rc_sync_p = -1;
+			return (NULL);
+		}
+	}
+
+	/*
+	 * Set up the cleanup function.  If none was specified, use a "no-op"
+	 * function (see the comment for cleanup_fake()),
+	 */
+	if (cleanup_routine == NULL)
+		cleanup_routine = cleanup_fake;
+	pthread_cleanup_push(cleanup_routine, user_cookie);
+
+	/*
+	 * Indicate that the main thread can continue (which will free U), and
+	 * run the desired routine. */
+	if (set_initialized(U) != 0) {
+		*U->rc_sync_p = -1;
+		rc = NULL;
+	} else {
+		/* Run the desired routine. */
+		rc = run_routine(user_cookie);
+	}
+
+	/* Clean up. */
+	pthread_cleanup_pop(1);
+
+	/* Done. */
+	return (rc);
+}
+
+/* Wait up 1 second, or until the condition variable is signalled. */
+static int
+cond_timedwait(struct wrapped_cookie * U)
+{
+	struct timespec tp = {0};
+	struct timeval tv;
+	int rc;
+
+	/* 1 second later than the current time. */
+	if (monoclock_get(&tv)) {
+		warnp("monoclock_get");
+		goto err0;
+	}
+	tp.tv_sec = tv.tv_sec;
+	tp.tv_nsec = 1000 * tv.tv_usec;
+
+	/* Wait for up to 1 second. */
+	if ((rc = pthread_cond_timedwait(U->cond_p, U->mutex_p, &tp)) != 0) {
+		if (rc != ETIMEDOUT) {
+			warn0("pthread_cond_timedwait: %s", strerror(rc));
+			goto err0;
+		}
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failed! */
+	return (-1);
+}
+
+/**
+ * pthread_create_blocking(thread, attr, init_routine, run_routine,
+ *     cleanup_routine, arg):
+ * Run pthread_create() and block until the new ${thread} has started and
+ * ${init_routine} has finished.  ${init_routine} and ${cleanup_routine} may
+ * be NULL.  If ${init_routine} is not NULL and returns a non-zero value, do
+ * not attempt to run ${run_routine} and instead return that value.  When
+ * ${run_routine} finishes, make its returned value available via
+ * pthread_join().
+ */
+int
```

Comment:
Can we make `pthread_create_blocking` have the same return values as `pthread_create`, i.e. return an error number?

Normally in cpercivaware we use -1 to indicate failure but it seems that behaving like `pthread_create` should take priority here.

Reply:
Do you mean "if `pthread_create` fails, return that error number", or "if any `pthread_*` fails, return that error number?

I now have it returning a `pthread_create` failure, but if something like `pthread_mutex_lock` fails, it prints a warning and returns the generic -1.

Reply:
I'd say if any error occurs.  You can't return a generic -1 for some errors but not others; the caller should be treating nonzero returns as errno values, and that would result in `Operation not permitted` being bogusly printed.

Reply:
What should we do if

    int init_routine(void * cookie)

returns non-zero?

Currently, I'm returning that value from `pthread_create_blocking()`.  It seems useful to keep that functionality, but there could be confusion between the returned values being `errno` vs. being from the init.

The simplest solution I can think of is to suggest (in the API comment) that `init_routine` should return a negative value.  Then the calling code can treat rc < 0 as being from init, and rc > 0 as being errno.

Reply:
On second thought: why have an `init_routine` at all?  I mean, what's the use case for:

* not initializing stuff before calling `pthread_create_blocking()`, and
* not waiting until after `cleanup_routine()` has been set up.

... come to think of it, maybe we should move the `cleanup_routine` stuff away from `pthread_create_blocking.c` as well.  I mean, if I were designing an API from scratch then I'd rather have that as part of the "run a thread" thing, but given that `pthread_create()` doesn't have it, it's probably better to go for API compatibility since everybody [1] already knows what to expect from `pthread_create()`.

[1] everybody who's looking at the source code for cpercivaware, at least.

Reply:
Yeah, I'm nuking `init_routine` and `cleanup_routine`.  Simplifies a lot of things, and it allows a trivial diff when updating `pushbits.c` to use `_blocking()`.